### PR TITLE
Features/32 diff

### PR DESCRIPTION
**Pull request by coquelin77**
## Description

Please include a summary of the change and which issue/s is/are fixed. 
Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes: #32 

Changes proposed:
- added diff function

## Type of change

Please delete options that are not relevant.

- [X] New feature (non-breaking change which adds functionality)

Have you handled and tested all split configurations?
yes

### CODE REVIEW

File: heat/core/arithmetics.py [hunk 1]
```diff
@@ -63,6 +65,78 @@ def add(t1, t2):
     return operations.__binary_op(torch.add, t1, t2)
 
 
+def diff(a, n=1, axis=-1):
+    """
+    Calculate the n-th discrete difference along the given axis.
+    The first difference is given by out[i] = a[i+1] - a[i] along the given axis, higher differences are calculated by using diff recursively.
```
Comment (Markus-Goetz) [inline heat/core/arithmetics.py]: Shouldn't that be a[i + n] instead of a[i + 1]?

File: heat/core/arithmetics.py [hunk 2]
```diff
@@ -63,6 +65,78 @@ def add(t1, t2):
...
+    a : DNDarray
+        Input array
+    n : int, optional
+        The number of times values are differenced. If zero, the input is returned as-is.
+        n=2 is equivalent to ht.diff(ht.diff(a))
```
Comment (Markus-Goetz) [inline heat/core/arithmetics.py]: Default value is missing in description

File: heat/core/arithmetics.py (hunk 1 above)
Comment (Markus-Goetz) [inline heat/core/arithmetics.py]: Linebreak

File: heat/core/arithmetics.py [hunk 3]
```diff
@@ -63,6 +65,78 @@ def add(t1, t2):
...
+    axis : int, optional
+        The axis along which the difference is taken, default is the last axis.
+
+    diff : DNDarray
+        The n-th differences. The shape of the output is the same as a except along axis where the dimension is smaller by n.
```
Comment (Markus-Goetz) [inline heat/core/arithmetics.py]: Missing

```
Returns
----------
```

Linebreak

File: heat/core/arithmetics.py [hunk 4]
```diff
@@ -63,6 +65,78 @@ def add(t1, t2):
...
+        The type of the output is the same as the type of the difference between any two elements of a.
+        The split does not change. The outpot array is balanced.
+    """
+    if n == 0:
+        return a
+    if n < 0:
+        raise ValueError('diff requires that n be a positive number, got {}'.format(n))
+    if not isinstance(axis, int):
+        raise ValueError('axis must be an integer, is currently {}'.format(type(axis)))
```
Comment (Markus-Goetz) [inline heat/core/arithmetics.py]: This should be sanitize_axis()

File: heat/core/arithmetics.py [hunk 5]
```diff
@@ -63,6 +65,78 @@ def add(t1, t2):
...
+    if not isinstance(a, dndarray.DNDarray):
+        raise TypeError('\'a\' must be a DNDarray')
+    rank = a.comm.rank
+    size = a.comm.size
```
Comment (Markus-Goetz) [inline heat/core/arithmetics.py]: rank and size not needed for all if-branches

File: heat/core/arithmetics.py [hunk 6]
```diff
@@ -63,6 +65,78 @@ def add(t1, t2):
...
+    if not a.is_distributed():
+        ret = a.copy()
+        for _ in range(n):
+            axis_slice = [slice(None)] * len(ret.shape)
+            axis_slice[axis] = slice(1, None, None)
+            axis_slice_end = [slice(None)] * len(ret.shape)
+            axis_slice_end[axis] = slice(None, -1, None)
+            ret = ret[axis_slice] - ret[axis_slice_end]
+        return ret
```
Comment (Markus-Goetz) [inline heat/core/arithmetics.py]: this is an early-out, else branch can be simple dedented

File: heat/core/dndarray.py [hunk 7]
```diff
@@ -661,30 +661,34 @@ def balance_(self):
...
         keep_slice = [slice(None), ] * self.numdims
         # need to send from the last one with data
-        for spr in range(last_pr_w_data, first_pr_w_data - 1, -1):
-            if self.comm.rank == spr:
-                for pr in range(self.comm.size - 1, spr, -1):
-                    send_amt = abs((chunk_map[pr, self.split] - lshape_map[pr, self.split]).item())
-                    send_amt = send_amt if send_amt < self.lshape[self.split] else self.lshape[self.split]
-                    if send_amt:
+        # start from x then push the data to the next one. then do the same at x+1 until the last process
+        balanced_process = [False for _ in range(self.comm.size)]
```
Comment (Markus-Goetz) [inline heat/core/dndarray.py]: See bug fixes branch remarks

File: heat/core/tests/test_statistics.py [hunk 8]
```diff
@@ -279,6 +279,7 @@ def test_average(self):
 
         # check weighted average over all float elements of split 3d tensor, across split axis
         random_volume = ht.array(torch.randn((3, 3, 3), dtype=torch.float64), is_split=1)
+        # random_volume = ht.random.randn(3, 3, 3, dtype=ht.float64, split=)
```
Comment (Markus-Goetz) [inline heat/core/tests/test_statistics.py]: remove or use

File: heat/core/arithmetics.py (hunk 1 above)
Comment (coquelin77) [reply heat/core/arithmetics.py]: i thought so at first too. but numpy diff is just with the next element. the n is essentially just a recursion and not a difference in how far to go before taking the subtraction

**Pull request merged**